Modules/Core/Common/test/itkImageRegionIteratorGTest.cxx

```diff
@@ -0,0 +1,136 @@
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+// First include the header files to be tested:
+#include "itkImageConstIterator.h"
+#include "itkImageConstIteratorWithIndex.h"
+#include "itkImageConstIteratorWithOnlyIndex.h"
+#include "itkImageIterator.h"
+#include "itkImageIteratorWithIndex.h"
+#include "itkImageLinearConstIteratorWithIndex.h"
+#include "itkImageLinearIteratorWithIndex.h"
+#include "itkImageRegionConstIterator.h"
+#include "itkImageRegionConstIteratorWithIndex.h"
+#include "itkImageRegionConstIteratorWithOnlyIndex.h"
+#include "itkImageRegionExclusionConstIteratorWithIndex.h"
+#include "itkImageRegionExclusionIteratorWithIndex.h"
+#include "itkImageRegionIterator.h"
+#include "itkImageRegionIteratorWithIndex.h"
+#include "itkImageRegionReverseConstIterator.h"
+#include "itkImageRegionReverseIterator.h"
+#include "itkImageReverseConstIterator.h"
+#include "itkImageReverseIterator.h"
+#include "itkImageScanlineConstIterator.h"
+#include "itkImageScanlineIterator.h"
+#include "itkImageSliceConstIteratorWithIndex.h"
+#include "itkImageSliceIteratorWithIndex.h"
+
+#include "itkImage.h"
+#include <gtest/gtest.h>
+#include <type_traits> // For false_type and true_type.
+
+namespace
+{
+// Has_IsAtBegin tells whether the specified iterator type has an `IsAtBegin()` member function, returning a `bool`. It
+// is inspired by "How to detect whether there is a specific member variable in class?", answered by Andy Prowl, Jan 25,
+// 2013 at https://stackoverflow.com/a/14523787 (the `has_id` example).
+template <typename TIterator, typename = bool>
+struct Has_IsAtBegin : std::false_type
+{};
+
+template <typename TIterator>
+struct Has_IsAtBegin<TIterator, decltype(std::declval<TIterator>().IsAtBegin())> : std::true_type
+{};
+
+
+template <typename TIterator>
+void
+CheckConstructedAtBegin()
+{
+  using ImageType = typename TIterator::ImageType;
+  using IndexType = typename TIterator::IndexType;
+  using SizeType = typename TIterator::SizeType;
+  using RegionType = typename TIterator::RegionType;
+
+  const auto image = ImageType::New();
+
+  // Use a small image size, so that the unit test won't take a long time.
+  static constexpr itk::SizeValueType imageSizeValue{ 4 };
+
+  image->SetRegions(SizeType::Filled(imageSizeValue));
+  image->Allocate();
+
+  // Check various regions, specified by the following `indexValue` and `sizeValue` combinations:
+  for (const itk::IndexValueType indexValue : { 0, 1 })
+  {
+    for (const auto sizeValue : { itk::SizeValueType{ 1 }, imageSizeValue - 1 })
+    {
+      const RegionType imageRegion(IndexType::Filled(indexValue), SizeType::Filled(sizeValue));
+
+      const TIterator iterator(image, imageRegion);
+      TIterator       iteratorThatGoesToBegin = iterator;
+      iteratorThatGoesToBegin.GoToBegin();
+      EXPECT_EQ(iterator, iteratorThatGoesToBegin);
```

Comment:
@blowekamp  Re: https://github.com/InsightSoftwareConsortium/ITK/pull/4815#discussion_r1734596667
> However with the nested dispatching of the functions how does test report a failure? Does it indicate the type? Is some Information needed the the "EXPECT" statements?

Indeed, the "EXPECT" statements currently do not report which particular iterator type has failed, if there would be any failure at all. Would you like that? In practice, none of them fail, and they probably never will. So it might be a bit hypothetical. But we could add the name of the iterator type (just in case of a failure) as follows: 

```cpp
     EXPECT_EQ(iterator, iteratorThatGoesToBegin) << "IteratorType: " << typeid(TIterator).name();
```
In the rare (unlikely) case of such a test failure, it would then print, with the report of the test failure:

    IteratorType: class itk::ImageScanlineIterator<class itk::Image<double,3> >

Would you like that?

Reply:
Otherwise we could do a feature request to https://github.com/google/googletest to ask them to print the type names of the arguments, in case of a test failure! 😺